HACKING.adoc

```diff
@@ -302,6 +302,27 @@ If `boot/ocamlc` changes (e.g. because you ran `make bootstrap`), then
 the build will revert to the slower bytecode-compiled `ocamlc` until
 you do the above step again.
 
+=== Using merlin
+
+When building the compiler as described above, the cmi and cmt files that are
+produced are 99% of the time not readable by merlin, because it is usually
+updated to support new versions of the compiler after they get released.
+
+Therefore, to use merlin on the compiler, you want to build the compiler with an
+older version of itself. The easiest way to do this is to install the latest
+released version of the compiler, install dune, and run:
```

Comment:
If I understand correctly, you are papering over the fact that this only works when the compiler is buildable with the last released version (that is, when no bootstrap was in fact required since the last release). This will typically work well early in development cycles, and not at all late in development cycles. I think you should emphasize this, because as written the notes suggest that the method you propose should always work, while in fact we know it often does not. 

(Maybe one could mention the idea of developing compiler changes in the last release branch instead of trunk, as a way to avoid this problem and make it easy to test one's changes with the opam repository.)

Reply:
You do understand correctly, and I will add a note about that point, but I think you're painting a darker picture than necessary.
We do regularly need to bootstrap because of the way our build is setup: part of the distribution is built with the bootstrapped compiler, part with the compiler that was just built, and we're expecting the build artifacts of both compilers to be compatible. Which they regularly aren't.

But that does not mean that the distribution can't be fully built with an older compiler; for instance: `dune build @libs` on this branch from a 4.09 switch does work fine.

That being said: things do break sometimes, new features do end up being used in the compiler; and I will add some text about that.

Reply:
Ah, I hadn't realized the "compatible artifacts" thing (despite being the thing that is problematic with the no-flat-float-array kerkuffle right under my nose), thanks.